ocaml/typing/typecore.ml

```diff
@@ -5572,7 +5605,7 @@ and type_expect_
           None, expected_mode
       in
       let type_label_exp ((_, label, _) as x) =
-        check_construct_mutability label.lbl_mut argument_mode;
+        check_construct_mutability ~loc ~env label.lbl_mut argument_mode;
```

Comment:
Me and @lpw25 had a discussion again last week, the outcome of which has been incorprated to the document. To properly implement that would require #2344. Also, I think "mode vs. mutation" is orthogonal to this PR and should be fixed separately.

Of course those things are still in the future, and the question is whether or not it's safer/better to make the changes you suggested **_now_**, to which I think it's no. Currently there is just one modality `Global` which secretly entails `Many` and `Shared`. It would be messy to base any type-checking optimization on top of that.